### ISSUE TITLE
Suggestion for root CMakeLists.txt

### ISSUE THREAD

**Issue by Fabian188** [DrTimothyAldenDavis/SuiteSparse#175]
I propose a simple CMakeLists.txt

It makes use of the SuiteSparse CMake files via cmake external projects and just adds the root CMakeLists.txt.

This is just a template with a small selection of options to demonstrate how it could be re realized.

I'm far from a CMake guru but it appears as an appropriate way for a simple wrapper for the original SuiteSparse CMake structure.

Please feel free to use and modify the file as you like. I wrote it myself and as it is so small and easy I put it into public domain without any restrictions - no need to mention me. 

Example usage for an empty SuiteSparse folder: mkdir mybuild; cd mybuild; ccmake ..

When changing the CMakeLists.txt for some major changes it is necessary to clean the build directory (here mybuild) and call make clean of SuiteSparse. I testet it with dev2 from Nov 10 

[CMakeLists.txt](https://github.com/DrTimothyAldenDavis/SuiteSparse/files/9985263/CMakeLists.txt)

**Comment by DrTimothyAldenDavis** [comment]
Thanks -- I'll take a look.

**Comment by fabiencastan** [comment]
@Fabian188 Very useful! It works fine for me.
I just needed to remove all the lines with:
`INSTALL_COMMAND "" # if removed, tried to call make install in build`
If we set these lines, it disables the installation of the libraries, which is quite problematic!
If we remove these lines and run cmake with `cmake -DCMAKE_INSTALL_PREFIX=/custom/path`, it installs everything correctly in the custom path.

**Comment by fabiencastan** [comment]
Making a confusion in my various tests, I also needed to propagate the CMAKE_INSTALL_PREFIX:
```
set(CMAKE_ARGS
    -DNSTATIC:BOOL=(NOT BUILD_STATIC)
    -DCMAKE_INSTALL_PREFIX=${CMAKE_INSTALL_PREFIX}  # <<< added this line
    )
```

**Comment by Fabian188** [comment]
@fabiencastan thanks for the input. I just created a fork SuiteSparse-root-cmake and included your suggestion.

Feel free to suggest more :)

**Comment by Fabian188** [comment]
What I do not like with this approach is, that a "make" implies a "make install". This is probably due to the external projects.

**Comment by fabiencastan** [comment]
@Fabian188 Do you have a link to your proposal? Or could you create a draft PR?

**Comment by fabiencastan** [comment]
The link: https://github.com/Fabian188/SuiteSparse-root-cmake

**Comment by Fabian188** [comment]
@DrTimothyAldenDavis  I'm currently trying to fix this automatic make install issue. 
I want to add some more stuff, I need (only static libs, optionally use lib64 instead of lib) and plan to continue on the fork the next days.

**Comment by Fabian188** [comment]
Honestly, I beliefe a real central cmake configuration is a much better solution. 
* central configuration once and the cmake variables can be read everywhere directly
* e.g. finding and configuring BLAS and openMP only once
* integrate current projects via add_subdirectory() such that all shares a common cmake variable scope.
* my current proposal with ExternalProject_Add() does not really integrate

One could e.g. configure once and then select modules via ccmake or by make AMD, make CHOLMOD, ...

This would be much more work than my approach but in the end the solution is IMHO the better one with much less total cmake code. Central cmake configuration have been realized by https://github.com/jlblancoc/suitesparse-metis-for-windows and https://github.com/sergiud/SuiteSparse. We actually have a  own very old solution of my former colleague Simon Triebenbacher which consists of 20 CMakeLists.txt with a total of 700 lines, where the current SuiteSparse 6.0.0 has 35 CMakeLists.txt with a total of 20K lines. I'm happy to share Simon's outdated solution (he'll be happy, too).

**Comment by Fabian188** [comment]
I solved the instant make install. Now the procedure is as common:
* mkdir build; cd build
* ccmake ..
* make
* make install

**Comment by waebbl** [comment]
Not sure, if the use of `ExternalProject` is a good way to do this. From my understanding ExternalProject is meant to download, patch and build projects from **another** repository. The structure of the SuiteSparse repository rather guides me to think of the root directory as some kine of super project, which can either be build as a whole (with certain options for its sub projects). Alternatively the sub projects could be build independantly as far as the build- and run-time dependencies are met.
Wouldn't it be easier to just use `add_subdirectory` calls, eventually guarding the cmake code for the sub projects with a variable like `SUITESPARSE_IS_SUPER` to allow for code which might be different if built in the context of the whole project or as an independant sub project?

**Comment by Fabian188** [comment]
I also would prefer a central root CMakeLists.txt with add_subdirectory(). However the subprojects are currently independent projects with a total of 20K cmake lines and would require a lot cmake changes.

There are forks doing exactly this, is e.g. Sergui's at https://github.com/sergiud/SuiteSparse
But I'm not sure if these will be upgraded to SuiteSparse 6.0

**Comment by waebbl** [comment]
Thanks for the link, I wasn't aware of that fork. Should have read all comments before posting.

**Comment by DrTimothyAldenDavis** [comment]
I'm very reluctant to add a top-level CMakeLists.txt file.  It would essentially merge all of the suitesparse packages into a single package with a single build folder.  That would require lots of changes to each package cmake build system, and it would make suitesparse monolithic.

Currently, the suitesparse packages are a la carte.  End users and pick and choose which they want, and build each independently.  This is essential for licensing issues.  Each package has a different license -- some GPL, some LGPL, some Apache, some BSD, and one mixed (CHOLMOD: GPL and LGPL).  End users who can only use LGPL, Apache, and BSD, for example, can simply remove the GPL packages. 

Right now, the top-level Makefile is optional, and each package can just be built as its own package.

**Comment by waebbl** [comment]
> I'm very reluctant to add a top-level CMakeLists.txt file. It would essentially merge all of the suitesparse packages into a single package with a single build folder. That would require lots of changes to each package cmake build system, and it would make suitesparse monolithic.
> 
> Currently, the suitesparse packages are a la carte. End users and pick and choose which they want, and build each independently. This is essential for licensing issues. Each package has a different license -- some GPL, some LGPL, some Apache, some BSD, and one mixed (CHOLMOD: GPL and LGPL). End users who can only use LGPL, Apache, and BSD, for example, can simply remove the GPL packages.
> 
> Right now, the top-level Makefile is optional, and each package can just be built as its own package.

I don't think it would make suitesparse monolithic. You could add options to select each package individually and use the `cmake_dependent_option` to drive inter-package dependencies, so the top-level CMakeLists.txt just acts as a driver to select what is going to get build and also use this to control the licensing issues.

My view is from the perspective of a package maintainer for the source-based Gentoo Linux distribution. While the package is being released as a single tarball, we have to maintain several packages with different versioning, instead of a single package with options to build and install each of the suitesparse packages. It *could* make the life of maintainers easier if we would have the possibility of a top-level CMakeLists.txt.

**Comment by DrTimothyAldenDavis** [comment]
Thanks for the feedback.  I'll consider it in a future release (after I post v6.0.2, which should appear soon).

**Comment by DrTimothyAldenDavis** [comment]
v7.0.0.beta1 is now posted.  I'll be considering this for v7.1.0 in the near future.  I do agree it would make life easier for Windows users, as something to use in place of my simple (but linux/mac only) top-level Makefile.

**Comment by chengts95** [comment]
I have created a CMake file to build all modules independently. I don't think using ExternalProject violated any license since the codes are not even copied during the building stages.

**Comment by DrTimothyAldenDavis** [comment]
We have something like this in progress.  There are lots of pending changes to the SuiteSparse build system, in the various development branches.

**Comment by chengts95** [comment]
Thank you for the reply. What about the suggestions i made here #309 ? As I know, these `Find*.cmake` modules are supposed to be used for finding libraries not built by CMake. For CMake install targets we should use something like this: 
```CMake 
    install ( TARGETS amd_static
        EXPORT  ${PROJECT_NAME} #Export a target that can be install as SuiteSparse::amd_static
        ARCHIVE DESTINATION ${SUITESPARSE_LIBDIR} ) 
    target_include_directories(amd_static INTERFACE
        $<INSTALL_INTERFACE:${SUITESPARSE_INCLUDEDIR}>) #set the interface header dir
install (EXPORT  ${PROJECT_NAME}
        FILE  ${PROJECT_NAME}Config.cmake
        NAMESPACE SuiteSparse::
        DESTINATION ${CMAKE_INSTALL_PREFIX}/cmake) #install the cmake config modules 
```

**Comment by mmuetzel** [comment]
The CMake build rules are modified to export targets instead of modules now. (There can still be improvements to them. But the basic transition should be complete.)

If we'd like to replace the root Makefile by a root CMakeLists.txt, we'd probably need to add support for all existing Makefile targets.
IIUC, the `demos` target in the current root Makefile doesn't yet corresponds to CMake rules for all subprojects.
It would probably be good to convert those demos to ctests. (That would probably also make it easier to spot failing tests in the CI.)
I can try to start with that. But I'd probably need some help. For some libraries, I don't know how to correctly interpret the output of that target to the console.
E.g., does the following output in the CI mean that the test succeeded?
https://github.com/DrTimothyAldenDavis/SuiteSparse/actions/runs/6191999485/job/16811275748#step:9:1360
What would be the criteria to decide that the tests failed?

**Comment by Fabian188** [comment]
From my experience with ctest, there needs to be a tool doing the decision and returning either 0 (ok) or another value (at least return values >= 1 indicate errors).

I quickly checked cs_demo1/2/3.c and it seems that these always return 0 within main. 
So the demos shall probably be adapted for ctest if I understand them right.

If one does still want to run all tests and not exit on the first failing function, the tests could return an error value, main could sum the return values and return it for the demo. So only if all functions return 0, the result will be 0.

**Comment by DrTimothyAldenDavis** [comment]
> E.g., does the following output in the CI mean that the test succeeded?
> https://github.com/DrTimothyAldenDavis/SuiteSparse/actions/runs/6191999485/job/16811275748#step:9:1360
> What would be the criteria to decide that the tests failed?

Yes, those are all passing tests.  I would need to work through all my packages to revise the demos to return 0 from the main program, to make them suitable for ctest, however.  I am not sure if I do that now.

I've used ctest a bit, but not in my own package.  LAGraph is a package I contribute to quite a bit as a co-author, and it has a ctest (I didn't write the ctest framework, but I do write tests in LAGraph for ctest).  See https://github.com/GraphBLAS/LAGraph .  At some point (soon) when LAGraph becomes reasonably stable, I will mirror it inside SuiteSparse.  It's a package that relies on GraphBLAS to solve graph problems (or any library that implements the GraphBLAS C API, in particular, not just SuiteSparse/GraphBLAS).

A ctest framework would be a very good idea to add in the future.  It will take me a while to get to this though.

It wouldn't be necessary for a top-level cmake build system to enable all existing Makefile targets, at least at first.  ctest would be nice.  I also have a top-level "make cov" which does test coverage in all packages.  Those tests are substantial but only work on Linux.  Most work on the Mac I think.  I doubt any would work (as is) on Windows.  The GraphBLAS test coverage is a very strange one: it has to be run inside MATLAB, and it requires my own test coverage mechanism because you can't do statement coverage of a C code called by MATLAB using existing tools.  It takes about 2 hours to run.   I've run it on Mac and Linux; I don't think it works on Windows but it might.

**Comment by mmuetzel** [comment]
In the light of #383 and #408: Are SuiteSparse libraries meant to be compilable as stand-alone projects?
If they aren't and it would be ok to build all libraries via a root CMakeLists.txt (maybe with the option to select which libraries to build), that would probably make the design of the build system a bit easier.

**Comment by DrTimothyAldenDavis** [comment]
Some libraries do depend on internal include files that are not visible to any other library that uses SuiteSparse, like amd_internal.h.  The cmake for UMFPACK should find ../AMD/Include/amd_internal.h, shouldn't it?  Or is it now finding the installed AMD?

I think it would be OK to have a single root CMakeLists.txt that builds all the libraries, I think (with "add_subdirectory" I suppose?).

It would be important to have an option to select which libraries to build.  Some users can use BSD and LGPL codes but not GPL, for example.

**Comment by mmuetzel** [comment]
> Some libraries do depend on internal include files that are not visible to any other library that uses SuiteSparse, like amd_internal.h. The cmake for UMFPACK should find ../AMD/Include/amd_internal.h, shouldn't it? Or is it now finding the installed AMD?

If AMD wasn't built in the same source tree, its Config.cmake file was never generated.
If it was built in a different folder than the one selected by the Makefile and installed, that file would have been generated, but the internal header would not have been installed with it...
It's a bit strange that libraries depend on internal files of other libraries, but the build system gives users the freedom to build and install them as if they were independent projects (which they really aren't)...

**Comment by DrTimothyAldenDavis** [comment]
Got it.

I don't think I use internal files from one library in many places.  If it makes the cmake build system simpler, I could revise that, and simply place an exact copy of amd_internal.h into the UMFPACK/Source (alongside umfpack_internal.h).

I see I also do this in GPUQREngine, which includes SuiteSparseGPU_internal.hpp.

**Comment by mmuetzel** [comment]
> I see I also do this in GPUQREngine, which includes SuiteSparseGPU_internal.hpp.

And `spqrgpu.hpp` includes `GPUQREngine_SuiteSparse.hpp` which in turn includes `SuiteSparseGPU_internal.hpp`. So all of those headers would need to be copied as well to SPQR.

Alternatively, just keep the CUDA parts with the interdependency. And after 7.2.1 fixes the issue for the more commonly used libraries (i.e., UMFPACK), #385 could avoid that issue for the CUDA parts (maybe for 7.3.0).

**Comment by mmuetzel** [comment]
There is a root CMakeLists.txt file on the dev2 branch now.
Should this be closed as fixed? Or better wait until this is part of a release?

**Comment by Fabian188** [comment]
I suggest to keep up to release but I don't mind if anyone wants to close it now.

**Comment by DrTimothyAldenDavis** [comment]
Yes, let's keep it open until it's ready for a stable release, in case we find things in the meantime, or in a beta release.

**Comment by DrTimothyAldenDavis** [comment]
This is now fixed in SuiteSparse 7.4.0, in the default dev branch of the SuiteSparse repo.  I'll release a 7.4.0.beta1 soon.

Can you give it a try?

**Comment by Fabian188** [comment]
Thanks for the effort! It seems that only static build is now possible, what I very much appreciate!

Is it ok to report issues here, or do you want separate issues?

**Comment by Fabian188** [comment]
I use `Apple clang version 15.0.0 (clang-1500.0.40.1) Target: arm64-apple-darwin23.1.0`
This error is a classic
```
file included from /Users/fwein/code/SuiteSparse-org/CHOLMOD/Supernodal/cholmod_super_numeric.c:88:
/Users/fwein/code/SuiteSparse-org/CHOLMOD/Supernodal/t_cholmod_super_numeric_worker.c:777:17: error: call to undeclared function 'SUITESPARSE_BLAS_ssyrk'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
                SUITESPARSE_BLAS_ssyrk ("L", "N",
                ^
```
The solution is simpy to add `-Wno-implicit-function-declaration` to CMAKE_C_FLAGS.

**Comment by Fabian188** [comment]
There are some warnings of the following type which you might want to check, looks like B->dtype is meant?
```
/Users/fwein/code/SuiteSparse-org/CHOLMOD/MatrixOps/cholmod_horzcat.c:92:56: warning: self-comparison always evaluates to false [-Wtautological-compare]
    if (mode != 0 && (A->xtype != B->xtype || A->dtype != A->dtype))

```

**Comment by Fabian188** [comment]
I have the following linker error

```
[  9%] Linking C executable cholmod_di_demo
ld: warning: ignoring duplicate libraries: '-lemutls_w', '-lgcc', '../SuiteSparse_config/libsuitesparseconfig.a'
ld: Undefined symbols:
  _SUITESPARSE_BLAS_cgemm, referenced from:
      _cs_cholmod_super_numeric_worker in libcholmod.a[70](cholmod_super_numeric.c.o)
      _zs_cholmod_super_numeric_worker in libcholmod.a[70](cholmod_super_numeric.c.o)
      _cs_cholmod_super_lsolve_worker in libcholmod.a[71](cholmod_super_solve.c.o)
      _cs_cholmod_super_ltsolve_worker in libcholmod.a[71](cholmod_super_solve.c.o)
  _SUITESPARSE_BLAS_cgemv, referenced from:
```

My BLAS settings are 
```
CMakeCache.txt:ALLOW_64BIT_BLAS:BOOL=OFF
CMakeCache.txt:BLAS_Accelerate_LIBRARY:FILEPATH=/Library/Developer/CommandLineTools/SDKs/MacOSX14.0.sdk/System/Library/Frameworks/Accelerate.framework
CMakeCache.txt:BLAS_mkl_intel_lp64_LIBRARY:FILEPATH=BLAS_mkl_intel_lp64_LIBRARY-NOTFOUND
...
```

**Comment by Fabian188** [comment]
I do automated lib building via cmake ExternalProjects. Therefore I would prefer to have the option not to build tests and demos. I have 
`CMakeCache.txt:DEMO:BOOL=OFF`
but `cholmod_di_demo` is still built.

**Comment by DrTimothyAldenDavis** [comment]
For the SUITESPARSE_BLAS_* errors:  those are new macros defined in SuiteSparse_config.h.  Can you try a clean build?  You must be getting a stale SuiteSparse_config.h from another source, or from a stale build perhaps.  The  older version of SuiteSparse that doesn't have those macros but they now appear here:

https://github.com/DrTimothyAldenDavis/SuiteSparse/blob/28951299a298f0fb28ddd625ce881d246344a457/SuiteSparse_config/SuiteSparse_config.h#L1239

For cholmod_di_demo:  that's used for the ctests by CMake.  It's not just a demo, so it's not disabled if DEMO is off.  I have four places in cmake where enable_testing is used:  the top-level CMakeLists.txt, and the CMakeLists.txt in CHOLMOD, LAGraph, and Mongoose.  I don't have a flag that disables testing but I could consider adding one.

The cholmod_horzcat is a bug -- thanks for catching that!  I just pushed a change to the dev2 branch and will add this to SuiteSparse 7.4.0.beta once it's ready to post.

**Comment by mmuetzel** [comment]
> The cholmod_horzcat is a bug -- thanks for catching that! I just pushed a change to the dev2 branch and will add this to SuiteSparse 7.4.0.beta once it's ready to post.

I'm not sure what happens with comments on commits. Do you receive notifications for those?
In case you do, sorry for the duplicate here:
That same change should probably also be done in `cholmod_vertcat.c`.

**Comment by Fabian188** [comment]
I have a complete empty build now. I build in parallel as GraphBlas takes quite some time

```
 92%] Building C object LAGraph/src/CMakeFiles/LAGraph.dir/algorithm/LAGr_Betweenness.c.o
[ 92%] Building C object LAGraph/src/CMakeFiles/LAGraph.dir/algorithm/LAGr_ConnectedComponents.c.o
In file included from In file included from In file included from /Users/fwein/code/SuiteSparse-org/LAGraph/src/algorithm/LAGr_BreadthFirstSearch.cIn file included from /Users/fwein/code/SuiteSparse-org/LAGraph/experimental/algorithm/LAGraph_AllKTruss.c/Users/fwein/code/SuiteSparse-org/LAGraph/src/algorithm/LAGr_BreadthFirstSearch.c:/Users/fwein/code/SuiteSparse-org/LAGraph/src/algorithm/LAGr_ConnectedComponents.c::In file included from /Users/fwein/code/SuiteSparse-org/LAGraph/experimental/algorithm/LAGraph_BF_basic.cIn file included from :/Users/fwein/code/SuiteSparse-org/LAGraph/src/algorithm/LAGr_Betweenness.cIn file included from :23:40:
In file included from In file included from /Users/fwein/code/SuiteSparse-org/LAGraph/src/algorithm/LG_alg_internal.h2664:
79:
:In file included from 26/Users/fwein/code/SuiteSparse-org/LAGraph/src/algorithm/LAGr_Betweenness.c/Users/fwein/code/SuiteSparse-org/LAGraph/src/utility/LG_internal.h::
:
:In file included from In file included from 29/Users/fwein/code/SuiteSparse-org/LAGraph/src/algorithm/LG_alg_internal.h/Users/fwein/code/SuiteSparse-org/LAGraph/src/utility/LG_internal.h79In file included from ::
/Users/fwein/code/SuiteSparse-org/LAGraph/src/utility/LG_internal.h:
In file included from :In file included from 2429/Users/fwein/code/SuiteSparse-org/LAGraph/experimental/algorithm/LAGraph_AllKCore.c::
:32:
24In file included from /Users/fwein/code/SuiteSparse-org/LAGraph/src/utility/LG_internal.h/Users/fwein/code/SuiteSparse-org/LAGraph/src/algorithm/LG_alg_internal.h:
::
In file included from /Users/fwein/code/SuiteSparse-org/LAGraph/src/utility/LG_internal.h:In file included from 29:
/Users/fwein/code/SuiteSparse-org/LAGraph/src/utility/LG_internal.h29::
29/Users/fwein/code/SuiteSparse-org/LAGraph/src/utility/LG_internal.h29::
:/Users/fwein/code/SuiteSparse-org/LAGraph/include/LAGraph.h/Users/fwein/code/SuiteSparse-org/LAGraph/include/LAGraph.h24/Users/fwein/code/SuiteSparse-org/LAGraph/include/LAGraph.h:29:
/Users/fwein/code/SuiteSparse-org/LAGraph/include/LAGraph.h66:
/Users/fwein/code/SuiteSparse-org/LAGraph/include/LAGraph.h:
:In file included from 66:/Users/fwein/code/SuiteSparse-org/LAGraph/include/LAGraph.h6:::: 66/Users/fwein/code/SuiteSparse-org/LAGraph/src/utility/LG_internal.h::666629::66::::
66 :::
6 error: 6:/Users/fwein/code/SuiteSparse-org/LAGraph/include/LAGraph.h:6 error: ::66 "The GraphBLAS library must support the v2.0 C API Specification"
/Users/fwein/code/SuiteSparse-org/LAGraph/include/LAGraph.h"The GraphBLAS library must support the v2.0 C API Specification"error:     #error "The GraphBLAS library must support the v2.0 C API Specification":error: 
"The GraphBLAS library must support the v2.0 C API Specification":
66error: :6:  6    #error "The GraphBLAS library must support the v2.0 C API Specification"
     ^


```

**Comment by Fabian188** [comment]
Many warnings of the following type

```
/Users/fwein/code/SuiteSparse-org/LAGraph/src/algorithm/LAGr_Betweenness.c:312:14: warning: due to lvalue conversion of the controlling expression, association of type 'const float' will never be selected because it is qualified [-Wunreachable-code-generic-assoc]
/usr/local/include/GraphBLAS.h:5639:17: note: expanded from macro 'GrB_assign'
                GB_CASES (, GrB, Matrix_assign) ,               \
                ^
/usr/local/include/GraphBLAS.h:665:15: note: expanded from macro 'GB_CASES'
        const float      p : GB_CONCAT ( prefix, _, func, _FP32   ),    \
              ^

```

**Comment by Fabian188** [comment]
>  I don't have a flag that disables testing but I could consider adding one.

I would appreciate it. I often patch packages to add this feature. We build all dependencies quite often in the build pipeline of our software and I guess many projects do like this (e.g. ParaView, but not for suitesparse). Building libs faster saves CO2 :)

**Comment by Fabian188** [comment]
Probably a follow up. Many errors of type

```
/Users/fwein/code/SuiteSparse-org/LAGraph/include/LAGraph.h:702:5: error: unknown type name 'GrB_Scalar'; did you mean 'GxB_Scalar'?
    GrB_Scalar emin ;   ///< minimum edge weight: value, lower bound, or unknown
    ^~~~~~~~~~
    GxB_Scalar
/Users/fwein/code/SuiteSparse-org/LAGraph/include/LAGraph.h:702:5: error: unknown type name 'GrB_Scalar'; did you mean 'GxB_Scalar'?
/Users/fwein/code/SuiteSparse-org/LAGraph/include/LAGraph.h:702:5    GrB_Scalar emin ;   ///< minimum edge weight: value, lower bound, or unknown
:    ^~~~~~~~~~
    GxB_Scalar 

```

**Comment by Fabian188** [comment]
```
Users/fwein/code/SuiteSparse-org/LAGraph/experimental/algorithm/LAGraph_AllKCore.c:103:55: error: use of undeclared identifier 'GrB_VALUEEQ_INT64'
    GrB_IndexUnaryOp valueEQ = (maxDeg > INT32_MAX) ? GrB_VALUEEQ_INT64 : GrB_VALUEEQ_INT32 ;
                                                      ^
/Users/fwein/code/SuiteSparse-org/LAGraph/experimental/algorithm/LAGraph_AllKCore.c:103:75: error: use of undeclared identifier 'GrB_VALUEEQ_INT32'
    GrB_IndexUnaryOp valueEQ = (maxDeg > INT32_MAX) ? GrB_VALUEEQ_INT64 : GrB_VALUEEQ_INT32 ;
                                                                          ^
[ 92%] Building C object LAGraph/src/CMakeFiles/LAGraph_static.dir/algorithm/LAGr_ConnectedComponents.c.o
[ 92%] Building C object LAGraph/experimental/CMakeFiles/LAGraphX_static.dir/algorithm/LAGraph_BF_basic.c.o
[ 92%] Building C object LAGraph/experimental/CMakeFiles/LAGraphX_static.dir/algorithm/LAGraph_AllKTruss.c.o
/Users/fwein/code/SuiteSparse-org/LAGraph/experimental/algorithm/LAGraph_AllKCore.c:104:5: error: unknown type name 'GrB_IndexUnaryOp'; did you mean 'GrB_UnaryOp'?
    GrB_IndexUnaryOp valueLE = (maxDeg > INT32_MAX) ? GrB_VALUELE_INT64 : GrB_VALUELE_INT32 ;
    ^~~~~~~~~~~~~~~~
    GrB_UnaryOp
/usr/local/include/GraphBLAS.h:741:35: note: 'GrB_UnaryOp' declared here
typedef struct GB_UnaryOp_opaque *GrB_UnaryOp ;
                                  ^
[ 92%] Building C object LAGraph/experimental/CMakeFiles/LAGraphX_static.dir/algorithm/LAGraph_BF_basic_mxv.c.o

```

**Comment by Fabian188** [comment]
```
In file included from /Users/fwein/code/SuiteSparse-org/LAGraph/experimental/algorithm/LAGraph_BF_full1.c:75:
/Users/fwein/code/SuiteSparse-org/LAGraph/include/LAGraphX.h:160:16: error: use of undeclared identifier 'GxB_MAX_NAME_LEN'
    char name [LAGRAPH_MAX_NAME_LEN+4] ;
               ^
/Users/fwein/code/SuiteSparse-org/LAGraph/include/LAGraph.h:109:30: note: expanded from macro 'LAGRAPH_MAX_NAME_LEN'
#define LAGRAPH_MAX_NAME_LEN GxB_MAX_NAME_LEN
                             ^

```

**Comment by mmuetzel** [comment]
Could you please clarify what you mean by "I build in parallel"?
LAGraph depends on GraphBLAS. Building both in parallel likely leads to some mixup...

Are you using the root CMakeLists.txt? Ideally it would automagically build everything in the correct order...

**Comment by Fabian188** [comment]
> Are you using the root CMakeLists.txt? Ideally it would automagically build everything in the correct order...

In "build" I did cmake .., so yes. I agree, dependencies would help. serial GraphBLAS is very slow ...

**Comment by Fabian188** [comment]
After a "make clean" I still have

```
[ 92%] Building C object LAGraph/src/CMakeFiles/LAGraph.dir/algorithm/LAGr_Betweenness.c.o
In file included from /Users/fwein/code/SuiteSparse-org/LAGraph/src/algorithm/LAGr_Betweenness.c:79:
In file included from /Users/fwein/code/SuiteSparse-org/LAGraph/src/utility/LG_internal.h:29:
/Users/fwein/code/SuiteSparse-org/LAGraph/include/LAGraph.h:66:6: error: "The GraphBLAS library must support the v2.0 C API Specification"
    #error "The GraphBLAS library must support the v2.0 C API Specification"
     ^
/Users/fwein/code/SuiteSparse-org/LAGraph/include/LAGraph.h:702:5: error: unknown type name 'GrB_Scalar'; did you mean 'GxB_Scalar'?
    GrB_Scalar emin ;   ///< minimum edge weight: value, lower bound, or unknown
    ^~~~~~~~~~
    GxB_Scalar
/usr/local/include/GraphBLAS.h:1912:34: note: 'GxB_Scalar' declared here
typedef struct GB_Scalar_opaque *GxB_Scalar ;
                                 ^
In file included from /Users/fwein/code/SuiteSparse-org/LAGraph/src/algorithm/LAGr_Betweenness.c:79:
In file included from /Users/fwein/code/SuiteSparse-org/LAGraph/src/utility/LG_internal.h:29:

```

**Comment by mmuetzel** [comment]
> In "build" I did cmake .., so yes. I agree, dependencies would help. serial GraphBLAS is very slow ...

There are many directories called `build` in the SuiteSparse repository. To which one exactly are you referring?

**Comment by mmuetzel** [comment]
> I agree, dependencies would help. serial GraphBLAS is very slow ...

Could you please elaborate in which way dependencies would help? (Which dependencies?) What do you mean by "serial GraphBLAS"?

**Comment by mmuetzel** [comment]
> /usr/local/include/GraphBLAS.h:1912:34: note: 'GxB_Scalar' declared here

It looks like it picked up a GraphBLAS that is installed on your build system. Maybe, that version isn't new enough?
Does LAGraph build with the version of GraphBLAS in the SuiteSparse repository?

**Comment by Fabian188** [comment]
> There are many directories called build in the SuiteSparse repository. To which one exactly are you referring?

The one in the root `SuiteSparse-org/build` 
I test with my own build in the root

**Comment by mmuetzel** [comment]
> I test with my own build in the root

Sorry. I'm still not sure what you are doing. Could you please show the exact commands that you run in the order you run them?

**Comment by DrTimothyAldenDavis** [comment]
You're somehow getting a very old GraphBLAS.h.  Several years old.

**Comment by Fabian188** [comment]
I do within the root of SparseSuite:
```
mkdir mybuild
cd mybuild
cmake .. -DCMAKE_C_FLAGS="-Wno-implicit-function-declaration -Wno-unreachable-code-generic-assoc"
make
```
Edit: added the make after cmake.

With the default settings I have `CMakeCache.txt:USE_SYSTEM_GRAPHBLAS:BOOL=OFF`

**Comment by Fabian188** [comment]
I build GraphBLAS and with serial I mean I build in serial (not make -j 10)

**Comment by Fabian188** [comment]
This is with VERBOSE=1

```
/Library/Developer/CommandLineTools/usr/bin/make  -f LAGraph/src/CMakeFiles/LAGraph.dir/build.make LAGraph/src/CMakeFiles/LAGraph.dir/build
[ 92%] Building C object LAGraph/src/CMakeFiles/LAGraph.dir/algorithm/LAGr_Betweenness.c.o
cd /Users/fwein/code/SuiteSparse-org/mybuild/LAGraph/src && /Library/Developer/CommandLineTools/usr/bin/cc -DLAGraph_EXPORTS -DLG_DLL -DLG_LIBRARY -I/Users/fwein/code/SuiteSparse-org/mybuild/LAGraph/src -I/Users/fwein/code/SuiteSparse-org/LAGraph/src -I/Users/fwein/code/SuiteSparse-org/LAGraph/include -I/Users/fwein/code/SuiteSparse-org/LAGraph/src/utility -I/Users/fwein/code/SuiteSparse-org/LAGraph/test/include -I/usr/local/include -I/Users/fwein/code/SuiteSparse-org/GraphBLAS/Include -isystem /opt/homebrew/include -Wno-implicit-function-declaration -Wno-unreachable-code-generic-assoc -DLGDIR=/Users/fwein/code/SuiteSparse-org/LAGraph -O3 -DNDEBUG -O3 -DNDEBUG -std=gnu11 -arch arm64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX14.0.sdk -mmacosx-version-min=13.5 -fPIC -Xclang -fopenmp -MD -MT LAGraph/src/CMakeFiles/LAGraph.dir/algorithm/LAGr_Betweenness.c.o -MF CMakeFiles/LAGraph.dir/algorithm/LAGr_Betweenness.c.o.d -o CMakeFiles/LAGraph.dir/algorithm/LAGr_Betweenness.c.o -c /Users/fwein/code/SuiteSparse-org/LAGraph/src/algorithm/LAGr_Betweenness.c
In file included from /Users/fwein/code/SuiteSparse-org/LAGraph/src/algorithm/LAGr_Betweenness.c:79:
In file included from /Users/fwein/code/SuiteSparse-org/LAGraph/src/utility/LG_internal.h:29:
/Users/fwein/code/SuiteSparse-org/LAGraph/include/LAGraph.h:66:6: error: "The GraphBLAS library must support the v2.0 C API Specification"
    #error "The GraphBLAS library must support the v2.0 C API Specification"
     ^
/Users/fwein/code/SuiteSparse-org/LAGraph/include/LAGraph.h:702:5: error: unknown type name 'GrB_Scalar'; did you mean 'GxB_Scalar'?
    GrB_Scalar emin ;   ///< minimum edge weight: value, lower bound, or unknown
    ^~~~~~~~~~
    GxB_Scalar
/usr/local/include/GraphBLAS.h:1912:34: note: 'GxB_Scalar' declared here
typedef struct GB_Scalar_opaque *GxB_Scalar ;
                                 ^
In file included from /Users/fwein/code/SuiteSparse-org/LAGraph/src/algorithm/LAGr_Betweenness.c:79:
In file included from /Users/fwein/code/SuiteSparse-org/LAGraph/src/utility/LG_internal.h:29:
```

**Comment by Fabian188** [comment]
You are right, I have some old GraphBLAS.h in my system and the system include is used.

```
locate GraphBLAS.h
/opt/homebrew/Cellar/suite-sparse/7.0.1/include/GraphBLAS.h
/opt/homebrew/include/GraphBLAS.h
/usr/local/Cellar/suite-sparse/5.10.1_1/include/GraphBLAS.h
/usr/local/include/GraphBLAS.h
```
Probably because CMAKE_INSTALL_PREFIX is by default /usr/local and there is my old GraphBLAS.h

**Comment by DrTimothyAldenDavis** [comment]
/usr/local/include/GraphBLAS.h:1912:34: note: 'GxB_Scalar' declared here

That's the problem. You're getting a very old GraphBLAS.h in /usr/local/include. 

Many many errors would be expected

**Comment by mmuetzel** [comment]
> I do within the root of SparseSuite:
> ```
> mkdir mybuild
> cd mybuild
> cmake .. -DCMAKE_C_FLAGS="-Wno-implicit-function-declaration -Wno-unreachable-code-generic-assoc"
> ```
> With the default settings I have `CMakeCache.txt:USE_SYSTEM_GRAPHBLAS:BOOL=OFF`

It looks like you are mixing CMake and make somehow.

After you've run the commands above, run `cmake --build .` without changing the folder.

**Comment by Fabian188** [comment]
I forgot the `make` after `cmake`. As I know the default generator, I usually do make (or ninja) directly.

**Comment by mmuetzel** [comment]
I believe I understand what might be happening. See #548.

I didn't notice that here because I was testing the root Makefile, the root CMakeLists.txt and compiling single projects with their Makefile and CMakeLists.txt (without cleaning up carefully in between).
So, for me it probably found the Config file in the GraphBLAS/build folder (which also works if it happens to exist).

Does the change from that PR fix it for you?

**Comment by mmuetzel** [comment]
> I don't have a flag that disables testing but I could consider adding one.

The flag `BUILD_TESTING` seems to be used quite commonly to configure whether a project should be built with its testing facility or not. We could probably do the same...

**Comment by Fabian188** [comment]
> Does the change from that PR fix it for you?
I applied it manually and without `make clean` it had no effect. I just wait for the effect with make clean.

Setting CMAKE_INSTALL_PREFIX to something other than /usr/local had no effect.

Next, I'll simply remove the old GraphBLAS.h in /usr/local/include but I see it likely that also on other macs brew puts it there by some dependency. Another order of the include dirs (build first, then system) might be the better approach.

**Comment by mmuetzel** [comment]
> I applied it manually and without `make clean` it had no effect. I just wait for the effect with make clean.

The Makefile probably needs to be regenerated for that change to have an effect (`make clean` probably won't help). If your CMake is new enough, you could use `cmake --fresh ..` for that: https://cmake.org/cmake/help/latest/manual/cmake.1.html#options

**Comment by mmuetzel** [comment]
> Next, I'll simply remove the old GraphBLAS.h in /usr/local/include but I see it likely that also on other macs brew puts it there by some dependency. Another order of the include dirs (build first, then system) might be the better approach.

I definitely agree that the build system should prefer its own libraries over the ones installed on the system. LAGraph has some special logic for that because one of the requirements was that it should be possible to build LAGraph with a different GraphBLAS.
I hope that #548 will make it so that GraphBLAS in SuiteSparse is preferred (if it is built).

**Comment by Fabian188** [comment]
Thanks for the hint with --fresh, I'll try after the current build.

Honestly I did not find a way yet to enable or disable building something, e.g. GraphBLAS.

**Comment by mmuetzel** [comment]
It's described in the `README.md`. Quoting here for convenience:

> * `SUITESPARSE_ENABLE_PROJECTS`:
> 
>   Semicolon separated list of projects to be built or `all`.
>   Default: `all` in which case the following projects are built:
> 
>   `suitesparse_config;mongoose;amd;btf;camd;ccolamd;colamd;cholmod;cxsparse;ldl;klu;umfpack;paru;rbio;spqr;spex;graphblas;lagraph`
> 
>   Additionally, `csparse` can be included in that list to build CSparse.
> 

If you like to disable building GraphBLAS and have the build system pick up a version installed on the system, you could configure with `cmake -DUSE_SYSTEM_GRAPHBLAS=ON`.
Documentation for that is a little lower in the same file:

> * `USE_SYSTEM_GRAPHBLAS`:
> 
>   If `ON`, use GraphBLAS libraries installed on the build system. If `OFF`,
>   automatically build GraphBLAS as dependency if needed. Default: `OFF`.
>

**Comment by Fabian188** [comment]
Some critical comments on the cmake:

There is GRAPHBLAS_BUILD_STATIC_LIBS and BUILD_STATIC_LIBS.

I do not see, for what I shall be able to edit the major/minor/sub of the packages. I would hide these variables by mark_as_advanced(<VAR>)

<img width="750" alt="Bildschirmfoto 2023-11-29 um 17 40 53" src="https://github.com/DrTimothyAldenDavis/SuiteSparse/assets/33481114/47ba9c6a-2c71-4d35-9dc8-a49738456635">

**Comment by mmuetzel** [comment]
> There is GRAPHBLAS_BUILD_STATIC_LIBS and BUILD_STATIC_LIBS.

Building GraphBLAS takes pretty long. So, there is a separate flag to *really* enable building static GraphBLAS libraries. `BUILD_STATIC_LIBS` alone won't do it. (That's also documented in the README.md.)

**Comment by mmuetzel** [comment]
> I do not see, for what I shall be able to edit the major/minor/sub of the packages. I would hide these variables by mark_as_advanced()

Thank you for the suggestion. We might do that at some point in the future. (But that's not a regression with the new root CMakeLists.txt afaict.)

**Comment by Fabian188** [comment]
One could at least have SUITESPARSE_ENABLE_PROJECTS with its default as option in the root CMakeLists.txt

Also cmake --fresh and a set CMAKE_INSTALL_PREFIX did not help (with the patch manually applied). When I remove the /usr/local/include/GraphBLAS.h, all builds with the default configuration + CMAKE_C_FLAGS="-Wno-implicit-function-declaration -Wno-unreachable-code-generic-assoc" on macOS Sonoma with Apple's clang 15.

I suggest to add this via
```
if(CMAKE_CXX_COMPILER_ID MATCHES "Clang" AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 15)
 set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wno-implicit-function-declaration -Wno-unreachable-code-generic-assoc")
endif()
```
As clang on Linux usually has the same issues. Similar for IntelLLVM which is clang but not identified by "Clang".


Do you want me to test on Windows with MSVC or Intel's oneAPI 2023.2, too? Not today.

**Comment by DrTimothyAldenDavis** [comment]
Is the LAGraph issue fixed for you now?

**Comment by DrTimothyAldenDavis** [comment]
In the SuiteSparse dev2 branch, I mean.

**Comment by mmuetzel** [comment]
> CMAKE_C_FLAGS="-Wno-implicit-function-declaration -Wno-unreachable-code-generic-assoc" on macOS Sonoma with Apple's clang 15.

Do you actually need those flags? Or did you only use them to work around an issue when there was still an older SuiteSparse_config.h in your build or source tree?

**Comment by mmuetzel** [comment]
> One could at least have SUITESPARSE_ENABLE_PROJECTS with its default as option in the root CMakeLists.txt

Could you please elaborate? It is in the root CMakeLists.txt:
https://github.com/DrTimothyAldenDavis/SuiteSparse/blob/28951299a298f0fb28ddd625ce881d246344a457/CMakeLists.txt#L32-L44

**Comment by Fabian188** [comment]
> > CMAKE_C_FLAGS="-Wno-implicit-function-declaration -Wno-unreachable-code-generic-assoc" on macOS Sonoma with Apple's clang 15.
> 
> Do you actually need those flags? Or did you only use them to work around an issue when there was still an older SuiteSparse_config.h in your build or source tree?

Compiling breaks without -Wno-implicit-function-declaration. Independent on any build configuration, that's simply clang >= 15

**Comment by Fabian188** [comment]
> Could you please elaborate? It is in the root CMakeLists.txt

I would expect it as option to be configured by ccmake. I usually do not modify a CMakeLists.txt but configure it.

There are already options, e.g.

`option ( ENABLE_CUDA "Enable CUDA acceleration" ON )`

**Comment by Fabian188** [comment]
Sorry, I just see, that it is already there in SUITESPARSE_ENABLE_PROJECTS
I overlooked it - my fault.

One could replace the default all with the content of SUITESPARSE_ALL_PROJECTS, then the user might easily delete what is not wanted or set `all` manually

**Comment by mmuetzel** [comment]
> Compiling breaks without -Wno-implicit-function-declaration. Independent on any build configuration, that's simply clang >= 15

I'm building with Clang 17 and I don't see that problem.
Can you reproduce the error without that flag in a clean source and build tree?

**Comment by mmuetzel** [comment]
The problem with the GraphBLAS header might actually be caused by this line:
https://github.com/DrTimothyAldenDavis/SuiteSparse/blob/da9f318227fabb989b396f6c3f8bc7ba062fc07d/LAGraph/CMakeLists.txt#L198

That looks odd to have in a (portable) CMake file. Is there a reason this was needed?

**Comment by DrTimothyAldenDavis** [comment]
That's a holdover from the original LAGraph repo: 

https://github.com/GraphBLAS/LAGraph/blob/7887f54875d5659e701809d623031fe0afd0aa0c/CMakeLists.txt#L165

We should probably remove it.

**Comment by DrTimothyAldenDavis** [comment]
I didn't write the original LAGraph CMakeLists.txt.  I can't recall why it got added.  I think it's an attempt to pick up a non-SuiteSparse GraphBLAS, but that should be handled by FindGraphBLAS.cmake anyway (if there is another version).

So let's delete that line.

**Comment by DrTimothyAldenDavis** [comment]
The original reason for that line is here, in the comments:

https://github.com/GraphBLAS/LAGraph/blob/09c7d372374178c1370b2e8c3ae8bb2cfbb2c939/CMakeLists.txt#L170

**Comment by mmuetzel** [comment]
I opened #552 that deletes that line.

**Comment by DrTimothyAldenDavis** [comment]
I'll have a few more changes as I sync the SuiteSparse/LAGraph v1.1 with the original LAGraph repo.  So far it's looking good; just one tiny change to the LAGraph/CMakelists.txt.  I'll push that to SuiteSparse shortly.

**Comment by DrTimothyAldenDavis** [comment]
These changess allow LAGraph to use GraphBLAS 7.1.0 (using its old build system, and older style FindGraphBLAS.cmake), and it allows LAGraph to be built outside of SuiteSparse:

https://github.com/DrTimothyAldenDavis/SuiteSparse/commit/5e4e5df0725ea033d1e6a498c11dcf19fe5e4850

The LAGraph/CMakeLists.txt file doesn't change much.  Just the place where it doesn't find itself inside SuiteSparse.  I had to copy the SuiteSparsePolicy.cmake into the LAGraph/cmake_modules, which is a bit unfortunate, but it works.

**Comment by Fabian188** [comment]
I freshly cloned 
```
git clone https://github.com/DrTimothyAldenDavis/SuiteSparse.git SuiteSparse-org
cd SuiteSparse-org
git checkout dev2
git log | head -> allow LAGraph to be used stand-alone, outside of SuiteSparse
mkdir mybuild
cd mybuild
cmake ..
make
```
And compilation fails
```
[  7%] Building C object CHOLMOD/CMakeFiles/CHOLMOD_static.dir/Supernodal/cholmod_l_super_numeric.c.o
In file included from /Users/fwein/code/SuiteSparse-org/CHOLMOD/Supernodal/cholmod_l_super_numeric.c:13:
In file included from /Users/fwein/code/SuiteSparse-org/CHOLMOD/Supernodal/cholmod_super_numeric.c:88:
/Users/fwein/code/SuiteSparse-org/CHOLMOD/Supernodal/t_cholmod_super_numeric_worker.c:777:17: error: call to undeclared function 'SUITESPARSE_BLAS_ssyrk'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
                SUITESPARSE_BLAS_ssyrk ("L", "N",
```
Interesting that it does fail for Apple clang version 15.0.0 but not vanialla clang 17, often I have similar issues on both - for IntelLLVM (also clang based) the issues are more often different. 

I suggest to handle this for "AppleClang".

**Comment by mmuetzel** [comment]
Does the following change fix the error for you?
``` diff
diff --git a/CHOLMOD/Supernodal/t_cholmod_super_numeric_worker.c b/CHOLMOD/Supernodal/t_cholmod_super_numeric_worker.c
index 8cec363cb..fdfee8f29 100644
--- a/CHOLMOD/Supernodal/t_cholmod_super_numeric_worker.c
+++ b/CHOLMOD/Supernodal/t_cholmod_super_numeric_worker.c
@@ -9,6 +9,7 @@
 //------------------------------------------------------------------------------
 
 #include "cholmod_template.h"
+#include "SuiteSparse_config.h"
 
 // Template routine for cholmod_super_numeric.  All xtypes supported, except
 // that a zomplex A and F result in a complex L (there is no supernodal zomplex
```

**Comment by Fabian188** [comment]
> Does the following change fix the error for you?

No, same error.

**Comment by DrTimothyAldenDavis** [comment]
This is the same problem:  you are getting a stale SuiteSparse_config.h.

The old SuiteSparse_config.h does not have macros that wrap the single precision BLAS.  THe new one does.  So you're getting an old SuiteSparse_config.h from somewhere.

**Comment by mmuetzel** [comment]
Could you please run `cmake --fresh .. | tee cmake_config.log` and upload that file here?

**Comment by DrTimothyAldenDavis** [comment]
I should revise all of my public headers to make them check version numbers of my other packages they rely on. They would #error in the C/C++ preprocessor if the version is outside a valid range. The error message would be more clear than something like "Suitesparse BLAS ssyrk not defined" from the compiler. 

All my packages have #defined package_VERSION_* macros just for this purpose. 

For example, cholmod 5.1.0 requires suitesparse_config 7.4.0 and will die badly (what you see here) if it finds an older Suitesparse_config.h.

**Comment by Fabian188** [comment]
> Could you please run `cmake --fresh .. | tee cmake_config.log` and upload that file here?

Here you are - I'm quite busy today and almost not at my computer

[cmake_config.log](https://github.com/DrTimothyAldenDavis/SuiteSparse/files/13515119/cmake_config.log)

**Comment by Fabian188** [comment]
> This is the same problem: you are getting a stale SuiteSparse_config.h.
> 
> The old SuiteSparse_config.h does not have macros that wrap the single precision BLAS. THe new one does. So you're getting an old SuiteSparse_config.h from somewhere.

You are right, there is a SuiteSparse_config.h in my /usr/local/include from nov 2022. 
What issue do you claim to come from that? The -Wimplicit-function-declaration thing?
```
/Users/fwein/code/SuiteSparse-org/CHOLMOD/Supernodal/t_cholmod_super_numeric_worker.c:777:17: error: call to undeclared function 'SUITESPARSE_BLAS_ssyrk'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
                SUITESPARSE_BLAS_ssyrk ("L", "N",
```
I still have it after removing the file in /usr/local/include and cmake --fresh. It compiles with the suppression of the warning.

**Comment by mmuetzel** [comment]
It looks like OpenMP is adding that path to the include directories:
```
-- OpenMP C libraries:      /opt/homebrew/lib/libomp.dylib 
-- OpenMP C include:        /opt/homebrew/include 
-- OpenMP C flags:          -Xclang -fopenmp 
```
I'll try and propose some changes.
@Fabian188: If it's ok with you, I'll ping you if I find something. Could you test changes when you come around to it?

Fwiw, it's the following for me (using Clang/MinGW):
```
-- OpenMP C libraries:      C:/msys64/clang64/lib/libomp.dll.a
-- OpenMP C include:
-- OpenMP C flags:          -fopenmp=libomp
```

**Comment by Fabian188** [comment]
> It looks like OpenMP is adding that to the include directories:
> 
> ```
> -- OpenMP C libraries:      /opt/homebrew/lib/libomp.dylib 
> -- OpenMP C include:        /opt/homebrew/include 
> -- OpenMP C flags:          -Xclang -fopenmp 
> ```

AAIK installing an external OpenMP is still the necessary variant for macOS. So all developers shall have this configuration.

But note, that the issues were on /usr/local/include.

brew is a common packager for macOS. With x86_64 it used /usr/local to install code. With arm64 the arm64 compiled tools are by default installed to /opt/homebrew. However one can run the x86_64 binaries from /usr/local transparently via emulation.

> 
> I'll try and propose some changes. @Fabian188: If it's ok with you, I'll ping you if I find something. Could you test changes when you come around to it?

Sure, I'll respond whenever it is possible. The question is no how to compile SuiteSparse out of the box on macOS, right?

**Comment by Fabian188** [comment]
in /opt/homebrew/include I again have the headers, now more recent

```
ll GraphBLAS.h SuiteSparse_config.h
lrwxr-xr-x  1 fwein  admin  48 28 Apr  2023 GraphBLAS.h -> ../Cellar/suite-sparse/7.0.1/include/GraphBLAS.h
lrwxr-xr-x  1 fwein  admin  57 28 Apr  2023 SuiteSparse_config.h -> ../Cellar/suite-sparse/7.0.1/include/SuiteSparse_config.h
include git:(master) pwd
/opt/homebrew/include

```

**Comment by mmuetzel** [comment]
> Sure, I'll respond whenever it is possible. The question is no how to compile SuiteSparse out of the box on macOS, right?

It compiles with Homebrew on macOS in the CI on GitHub. However, the CI doesn't install an older version of the SuiteSparse libraries before building. So, issues like the one you are seeing can go unnoticed...

**Comment by mmuetzel** [comment]
> in /opt/homebrew/include I again have the headers, now more recent

IIUC, you'd need the headers from version 7.4.0 (or newer). That (together with the OpenMP include directory) could explain the C99 error...
Maybe #555 could help with that.

**Comment by mmuetzel** [comment]
@DrTimothyAldenDavis: We can't do anything about older versions. But could the headers of the SuiteSparse libraries be installed in a sub-directory (e.g., `/usr/local/include/suitesparse`)?
Currently, there is the risk of headers from older libraries being picked up when the include path (e.g., `/usr/local/include`) gets added for other dependencies installed on the build system (e.g., OpenMP). That risk would be lower if the headers were installed in their own directory. CMake targets and pkg-config files would help dependent projects to find the correct folder with the headers.

Probably not something that should be changed for the current release (or ever). But maybe for a later release?

Edit: Looking at the commit log, this is probably not a new issue. But it might have gone unnoticed so far because there weren't any breaking changes in SuiteSparse_config.h in a long time...

**Comment by DrTimothyAldenDavis** [comment]
with this change to the dev2 branch:
https://github.com/DrTimothyAldenDavis/SuiteSparse/commit/bc21716ac739ad8827b3faa6b748a6f4153d57da
all headers are checked for the right versions, and an #error will occur in the compiler, if a stale header is found.

**Comment by DrTimothyAldenDavis** [comment]
> > This is the same problem: you are getting a stale SuiteSparse_config.h.
> > The old SuiteSparse_config.h does not have macros that wrap the single precision BLAS. THe new one does. So you're getting an old SuiteSparse_config.h from somewhere.
> 
> You are right, there is a SuiteSparse_config.h in my /usr/local/include from nov 2022. What issue do you claim to come from that? The -Wimplicit-function-declaration thing?
> 
> ```
> /Users/fwein/code/SuiteSparse-org/CHOLMOD/Supernodal/t_cholmod_super_numeric_worker.c:777:17: error: call to undeclared function 'SUITESPARSE_BLAS_ssyrk'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
>                 SUITESPARSE_BLAS_ssyrk ("L", "N",
> ```
> 
> I still have it after removing the file in /usr/local/include and cmake --fresh. It compiles with the suppression of the warning.

Yes, if the SUITESPARSE_BLAS_ssyrk is being seen as an undeclared function, then you're getting a stale SuiteSparse_config.h.  Their would be many other errors as well.

I have a stale SuiteSparse_config.h in /usr/local/include on my Mac, because that's where "make install" in SuiteSparse installs itself, and I had my own (old) copy there.  But doing a fresh build doesn't bring that in.  The old one is version 7.2.0, and the new one is 7.4.0, and none of my #error's trigger, at least with this update on the dev2 branch.

**Comment by DrTimothyAldenDavis** [comment]
The root CMakeLists.txt file is working well in the prerelease (SuiteSparse 7.4.0.beta1).  Shall I close this issue?

**Comment by Fabian188** [comment]
Thanks both of you very much for your effort! I close the issue. Possibly I suggest in the future variants of handling the root CMakeLists.txt.

**Comment by DrTimothyAldenDavis** [comment]
Thanks again for the suggestion and for the help in testing it.
